src/us/kbase/workspace/database/provenance/CreditMetadata.java

```diff
@@ -0,0 +1,465 @@
+				version);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		CreditMetadata other = (CreditMetadata) obj;
+		return Objects.equals(comments, other.comments)
+				&& Objects.equals(contributors, other.contributors)
+				&& Objects.equals(dates, other.dates)
+				&& Objects.equals(funding, other.funding)
+				&& Objects.equals(identifier, other.identifier)
+				&& Objects.equals(license, other.license)
+				&& Objects.equals(relatedIdentifiers, other.relatedIdentifiers)
+				&& Objects.equals(resourceType, other.resourceType)
+				&& Objects.equals(titles, other.titles)
+				&& Objects.equals(version, other.version);
+	}
+
+	/**
+	 * Gets a builder for the {@link CreditMetadata}.
+	 *
+	 * @param identifier
+	 *                unique persistent ID for the resource
+	 * @param contributors
+	 *                list of {@link Contributor} objects; at least one contributor is required
+	 * @param titles
+	 *                list of {@link Title} objects; at least one title is required
+	 * @return the builder.
+	 */
+	public static Builder getBuilder(
+			final String identifier,
+			final List<Contributor> contributors,
+			final List<Title> titles) {
+		return new Builder(identifier, contributors, titles);
+	}
+
+	/** A builder for {@link CreditMetadata}. */
+	public static class Builder {
+		private final String identifier;
+		private String license = null;
+		private String version = null;
+		private ResourceType resourceType = ResourceType.DATASET;
+		private List<String> comments = null;
+		private final List<Contributor> contributors;
+		private List<EventDate> dates = null;
+		private List<FundingReference> funding = null;
+		private List<PermanentID> relatedIdentifiers = null;
+		private final List<Title> titles;
+		private List<String> errorList = new ArrayList<>();
+
+		private Builder(
+				final String identifier,
+				final List<Contributor> contributors,
+				final List<Title> titles) {
+
+			String checkedPid = identifier;
+			try {
+				checkedPid = Common.checkPid(identifier, "identifier", false);
+			} catch (IllegalArgumentException e) {
+				this.errorList.add(e.getMessage());
+			}
+			this.identifier = checkedPid;
+
+			// must be at least one contributor
+			this.contributors = Common.dedupeSimpleList(contributors);
+			if (this.contributors == null || this.contributors.isEmpty()) {
+				this.errorList.add("at least one contributor must be provided");
+			}
+
+			// must be at least one title
+			this.titles = Common.dedupeSimpleList(titles);
+			if (this.titles == null || this.titles.isEmpty()) {
+				this.errorList.add("at least one title must be provided");
+			}
+		}
+
+		/**
+		 * Sets the resource type for the resource.
+		 *
+		 * As there is currently only one resource type, entering null will set the
+		 * resource type value to DEFAULT_RESOURCE_TYPE in the resultant object.
```

Comment:
Alternatively, you could just ignore null / whitespace only input as a noop rather than resetting to the default. In the current case, there would be no behavior change

Reply:
N/A

Reply:
:+1: